deisctl/cmd/upgrade.go

```diff
@@ -0,0 +1,110 @@
+package cmd
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/deis/deis/deisctl/backend"
+	"github.com/deis/deis/deisctl/etcdclient"
+)
+
+// UpgradePrep stops and uninstalls all components except router and publisher
+func UpgradePrep(b backend.Backend) error {
+	var wg sync.WaitGroup
+
+	b.Stop([]string{"database", "registry@*", "controller", "builder", "logger", "logspout"}, &wg, Stdout, Stderr)
+	wg.Wait()
+	b.Destroy([]string{"database", "registry@*", "controller", "builder", "logger", "logspout"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	b.Stop([]string{"store-volume", "store-gateway@*"}, &wg, Stdout, Stderr)
+	wg.Wait()
+	b.Destroy([]string{"store-volume", "store-gateway@*"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	b.Stop([]string{"store-metadata"}, &wg, Stdout, Stderr)
+	wg.Wait()
+	b.Destroy([]string{"store-metadata"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	b.Stop([]string{"store-daemon"}, &wg, Stdout, Stderr)
+	wg.Wait()
+	b.Destroy([]string{"store-daemon"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	b.Stop([]string{"store-monitor"}, &wg, Stdout, Stderr)
+	wg.Wait()
+	b.Destroy([]string{"store-monitor"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	fmt.Fprintln(Stdout, "The platform has been stopped, but applications are still serving traffic as normal.")
+	fmt.Fprintln(Stdout, "Your cluster is now ready for upgrade. Install a new deisctl version and run deisctl take-over.")
+	fmt.Fprintln(Stdout, "For more details, see: http://docs.deis.io/en/latest/managing_deis/upgrading-deis/#graceful-upgrade")
+	return nil
+}
+
+func listPublishedServices(etcdClient etcdclient.Client) ([]*etcdclient.ServiceKey, error) {
+	nodes, err := etcdClient.GetRecursive("deis/services")
+	if err != nil {
+		return nil, err
+	}
+
+	return nodes, nil
+}
+
+func republishServices(ttl uint64, nodes []*etcdclient.ServiceKey, etcdClient etcdclient.Client) error {
+	for _, node := range nodes {
+		_, err := etcdClient.Update(node.Key, node.Value, ttl)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UpgradeTakeover gracefully starts a platform stopped with UpgradePrep
+func UpgradeTakeover(b backend.Backend) error {
+	etcdClient, err := etcdclient.GetEtcdClient()
+	if err != nil {
+		return err
+	}
+
+	if err := doUpgradeTakeOver(b, etcdClient); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func doUpgradeTakeOver(b backend.Backend, etcdClient etcdclient.Client) error {
```

Comment:
@iancoffey should this function also update the value of `/deis/platform/version` in etcd?  Otherwise, won't the same versions of each component just come right back up (despite a newer version of deisctl having been used to do so)?

Referencing the inner-workings of the `get_image` function:

```
#!/bin/bash
# usage: get_image <component_path>
IMAGE=`etcdctl get $1/image 2>/dev/null`

# if no image was set in etcd, we use the default plus the release string
if [ $? -ne 0 ]; then
  RELEASE=`etcdctl get /deis/platform/version 2>/dev/null`

  # if no release was set in etcd, use the default provisioned with the server
  if [ $? -ne 0 ]; then
    source /etc/deis-release
    RELEASE=$DEIS_RELEASE
  fi

  IMAGE=$1:$RELEASE
fi

# remove leading slash
echo ${IMAGE#/}
```

Come to think of it, it needs to be noted that the upgrade won't upgrade any components for which a custom image has already been specified via etcd key `/deis/<component>/image`